[TensorExpr] Add support for constant tensors in tensorexpr kernel.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56319 [TensorExpr] Add support for constant tensors in tensorexpr kernel.**
* #56318 [TensorExpr] TensorExprKernel: switch type of tensors_ from Tensor to Buf.

With this change the TorchScript graph can have constant tensors in it
and we still will be able to lower it to TE. The constants are
registered (or bound) within the `TensorExprKernel` object and when the
codegen is called, they are passed along with usual inputs and outputs.

Differential Revision: [D27838747](https://our.internmc.facebook.com/intern/diff/D27838747)

Commit:
[TensorExpr] Add support for constant tensors in tensorexpr kernel.

With this change the TorchScript graph can have constant tensors in it
and we still will be able to lower it to TE. The constants are
registered (or bound) within the `TensorExprKernel` object and when the
codegen is called, they are passed along with usual inputs and outputs.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add support for constant tensors in tensorexpr kernel."


With this change the TorchScript graph can have constant tensors in it
and we still will be able to lower it to TE. The constants are
registered (or bound) within the `TensorExprKernel` object and when the
codegen is called, they are passed along with usual inputs and outputs.

Differential Revision: [D27838747](https://our.internmc.facebook.com/intern/diff/D27838747)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add support for constant tensors in tensorexpr kernel."


With this change the TorchScript graph can have constant tensors in it
and we still will be able to lower it to TE. The constants are
registered (or bound) within the `TensorExprKernel` object and when the
codegen is called, they are passed along with usual inputs and outputs.

Differential Revision: [D27838747](https://our.internmc.facebook.com/intern/diff/D27838747)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -211,6 +218,9 @@ class TORCH_API TensorExprKernel {
       known_sizes_;
+
+  std::vector<at::Tensor> unpacked_constant_tensors_;
+  std::vector<ConstantDescr> constants_;
```
Do we prefer lowerCamelCase or lower_under_scores?  The existing members go both ways which makes me (╯°□°)╯︵ ┻━┻

ZolotukhinM replied:
Agreed, maybe it's time to do some spring cleaning :) I've found pytorch cpp guidelines (unfortunately, the link is FB-internal): https://www.internalfb.com/intern/wiki/PyTorch/PyTorchDev/Workflow/Style/

The relevant part is this:

> NAMING
> * We follow the Facebook C++ Style Guidelines. Examples: some_namespace, SomeClass, SomeEnum, someFunction, someMethod, someMemberVariable, somePrivateMemberVariable_, SOME_MACRO
> * Exception: C++ methods and functions bound into python use python naming scheme (snake_case, ...) also in their C++ names. Exposing a previously unexposed method might need renaming the C++ method.
>

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2526,6 +2526,36 @@ Tensor* TensorExprKernel::convertOutputToCorrectStrides(torch::jit::Value* v) {
+
+  if (!const_tensor.is_contiguous()) {
+    const_tensor = const_tensor.clone().contiguous();
+    unpacked_constant_tensors_.push_back(const_tensor);
```
Something seems fishy here but I don't understand `at::Tensor` that well... if `const_tensor` isn't contiguous we `clone()` it and squirrel away the `data_ptr` in `constants_`... but then doesn't the `const_tensor.clone()` go out of scope, get deleted, and we have a dangling pointer?

Also, why do we need to store `unpacked_constant_tensors_`?  (I feel like these must be related but I don't understand how 😅 )

ZolotukhinM replied:
AFAIU, `at::Tensor` is a kind of shared ptr, so it could go out of the scope without bad consequences as long as there is at least one live pointer. The reason we're storing it to `unpacked_constant_tensors_` is exactly to keep it alive - this way the `TensorExprKernel` objects owns all constant tensors it materialized via `contiguous` call.

bertmaher replied:
Oh this makes sense now.  I totally didn't read the code correctly, and now I see that you're storing `const_tensor.clone().contiguous()` back into `const_tensor`.  So now it's perfectly clear to me how the reference is kept alive.  Thanks!

ZolotukhinM replied:
Yep, `at::Tensor` is just a ref-counted pointer to `TensorImpl`, see https://github.com/pytorch/pytorch/blob/7636cb6bab1c7ac2c0285935793404bef3bb546a/aten/src/ATen/templates/TensorBody.h#L71-L88

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2526,6 +2526,36 @@ Tensor* TensorExprKernel::convertOutputToCorrectStrides(torch::jit::Value* v) {
+    // into immediates in TE IR
+    return;
+  }
+  auto n = v->node();
```
Man, this time clang-time is on point! :)

ZolotukhinM commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -1125,5 +1126,72 @@ TEST_F(Kernel, SanitizeNames_CUDA) {
+  TensorExprKernel k(graph);
+  auto stmt = k.getCodeGenStmt();
+  std::ostringstream oss;
+  oss << *stmt;
```
Ah, that's a leftover from a test I copied that from, I'll cleanup!

Commit:
Update on "[TensorExpr] Add support for constant tensors in tensorexpr kernel."


With this change the TorchScript graph can have constant tensors in it
and we still will be able to lower it to TE. The constants are
registered (or bound) within the `TensorExprKernel` object and when the
codegen is called, they are passed along with usual inputs and outputs.

Differential Revision: [D27838747](https://our.internmc.facebook.com/intern/diff/D27838747)

[ghstack-poisoned]
